### PR TITLE
Fixes #5051 (introduced in #4684), effectively allowing underscores in address literals.

### LINKED ISSUE #5051: getChecksummedAddress crash with invalid input
Found by solfuzzer. Uploaded the example as it may contain unreadable characters.
[checksummed.txt](https://github.com/ethereum/solidity/files/2405615/checksummed.txt)

### PR DESCRIPTION
**Pull request by christianparpart**
The bug in #5051 was implicitly introduced in #4684, therefore no need for a Changelog entry.

I've added a small test for ensuring address literals with underscores do compile correctly. I am not sure though whether it's the best place (it was the best I found so far).

### CODE REVIEW

File: libsolidity/ast/AST.cpp [hunk 1]
```diff
@@ -648,20 +649,21 @@ bool Literal::looksLikeAddress() const
 	if (!isHexNumber())
 		return false;
 
-	return abs(int(value().length()) - 42) <= 1;
+	int const length = (int) count_if(next(next(begin(value()))), end(value()), [](char c) { return isxdigit(c); });
```
Comment (chriseth) [inline libsolidity/ast/AST.cpp]: Please add a comment why we do this and don't use c-style casts.

File: libsolidity/ast/AST.cpp [hunk 2]
```diff
@@ -648,20 +649,21 @@ bool Literal::looksLikeAddress() const
...
+	return abs(length - 40) <= 1;
```
Comment (chriseth) [inline libsolidity/ast/AST.cpp]: Might be easier to do something like `length >= 39 || length <= 41` - it does not need the cast from size_t to int.

**Review by chriseth — DISMISSED** [review]
Perhaps it would be better to introduce a function `valueWithoutUnderscores` into the Literal class?

File: libsolidity/ast/AST.cpp [hunk 3]
```diff
@@ -28,6 +28,7 @@
 
 #include <boost/algorithm/string.hpp>
 
+#include <cctype>
```
Comment (chriseth) [inline libsolidity/ast/AST.cpp]: Is this still needed?

File: libdevcore/CommonData.cpp [hunk 4]
```diff
@@ -93,6 +93,7 @@ bool dev::passesAddressChecksum(string const& _str, bool _strict)
 string dev::getChecksummedAddress(string const& _addr)
 {
 	string s = _addr.substr(0, 2) == "0x" ? _addr.substr(2) : _addr;
+	boost::erase_all(s, "_");
```
Comment (chriseth) [inline libdevcore/CommonData.cpp]: I have the feeling that CommonData.cpp should not be concerned about underscores. Can you move the "cleanup" to the call site in AST.cpp?

Comment (christianparpart) [reply libdevcore/CommonData.cpp]: I agree with that and I was close to removing it. Let me do that then.

**Review by chriseth — APPROVED** [review]

**Pull request merged**